streams/src/test/java/org/apache/kafka/streams/tests/RelationalSmokeTest.java

```diff
@@ -0,0 +1,962 @@
+                    );
+                    if (!passed) {
+                        LOG.info("Verification has not passed yet. ");
+                        Thread.sleep(500);
+                    }
+                }
+                return verifySync(
+                    true,
+                    consumedArticles,
+                    consumedComments,
+                    consumedAugmentedArticles,
+                    consumedAugmentedComments
+                );
+            }
+        }
+
+        public static <T> void assertThat(final AtomicBoolean pass,
+                                          final StringBuilder failures,
+                                          final String test,
+                                          final T actual,
+                                          final Matcher<? super T> matcher) {
+            if (!matcher.matches(actual)) {
+                if (failures != null) {
+                    final Description description = new StringDescription(failures);
+                    description.appendText("\n" + test)
+                               .appendText("\nExpected: ")
+                               .appendDescriptionOf(matcher)
+                               .appendText("\n     but: ");
+                    matcher.describeMismatch(actual, description);
+                    description.appendText("\n");
+                }
+                pass.set(false);
+            }
+        }
+
+        static boolean verifySync(final boolean logResults,
+                                  final Map<Integer, Article> consumedArticles,
+                                  final Map<Integer, Comment> consumedComments,
+                                  final Map<Integer, AugmentedArticle> consumedAugmentedArticles,
+                                  final Map<Integer, AugmentedComment> consumedAugmentedComments) {
+            final AtomicBoolean pass = new AtomicBoolean(true);
+            final StringBuilder report = logResults ? new StringBuilder() : null;
+
+            assertThat(pass, report, "one article", consumedArticles.size(), greaterThan(0));
+            assertThat(pass, report, "one comment", consumedComments.size(), greaterThan(0));
+
+            assertThat(
+                pass,
+                report,
+                "article size",
+                consumedAugmentedArticles.size(),
+                is(consumedArticles.size())
+            );
+            assertThat(
+                pass,
+                report,
+                "comment size",
+                consumedAugmentedComments.size(),
+                is(consumedComments.size())
+            );
+
+            final Map<Integer, Long> commentCounts = new TreeMap<>();
+
+            for (final RelationalSmokeTest.AugmentedComment augmentedComment : consumedAugmentedComments.values()) {
+                assertThat(pass, report, "comment missing", consumedComments, hasKey(augmentedComment.getKey()));
```

Comment:
Could we include specific article / comment information in the report as well like the line 894 below?

Also I'm wondering if multiple of the assertions have failed, would the later report replace the older one?

Reply:
Will do. No, they are just all concatenated in the report.